src/andromede/model/resolve_library.py

```diff
@@ -60,55 +60,100 @@ def resolve_library(
+                _update_treated_libs_and_import_stack(treated_lib_ids, import_stack)
+
+    return output_lib_dict
+
+
+def _add_preloaded_port_types_to_current_lib(
+    preloaded_port_types: dict[str, PortType], current_lib: Library
+) -> None:
+    current_lib.port_types.update(preloaded_port_types)
+
+
+def _add_treated_dependent_port_types_to_current_lib(
+    output_lib_dict: Dict[str, Library],
+    treated_lib_ids: Set[str],
+    cur_yaml_lib: InputLibrary,
+    current_lib: Library,
+) -> None:
+    done_dependencies = set(cur_yaml_lib.dependencies) & treated_lib_ids
+    for done_lib in done_dependencies:
+        current_lib.port_types.update(output_lib_dict[done_lib].port_types)
+
+
+def _update_treated_libs_and_import_stack(
+    treated_lib_ids: Set[str], import_stack: List[str]
+) -> None:
+    treated_lib_ids.add(import_stack.pop())
+
+
+def _treat_lib(
+    current_lib: Library, cur_yaml_lib: InputLibrary, output_lib: Dict[str, Library]
+) -> None:
+    port_types = [_convert_port_type(p) for p in cur_yaml_lib.port_types]
+    port_types_dict = dict((p.id, p) for p in port_types)
+
+    if current_lib.port_types.keys() & port_types_dict.keys():
+        raise Exception(
+            f"Port(s) : {str(current_lib.port_types.keys() & port_types_dict.keys())} is(are) defined twice."
+        )
+    current_lib.port_types.update(port_types_dict)
+
+    cur_yaml_lib_model_ids = [model.id for model in cur_yaml_lib.models]
+    for id in cur_yaml_lib_model_ids:
+        if cur_yaml_lib_model_ids.count(id) > 1:
+            raise Exception(f"Model {id} is defined twice")
+
+    models = [_resolve_model(m, current_lib.port_types) for m in cur_yaml_lib.models]
 
-                if output_lib.port_types.keys() & port_types_dict.keys():
-                    raise Exception(
-                        f"Port(s) : {str(output_lib.port_types.keys() & port_types_dict.keys())} is(are) defined twice."
-                    )
-                output_lib.port_types.update(port_types_dict)
+    models_dict = dict((m.id, m) for m in models)
 
-                models = [
-                    _resolve_model(m, output_lib.port_types) for m in cur_lib.models
-                ]
+    current_lib.models.update(models_dict)
+    output_lib[current_lib.id] = current_lib
 
-                models_dict = dict((m.id, m) for m in models)
-                if output_lib.models.keys() & models_dict.keys():
-                    raise Exception(
-                        f"Model(s) : {str(output_lib.models.keys() & models_dict.keys())} is(are) defined twice"
-                    )
-                output_lib.models.update(models_dict)
 
-                done.add(import_stack.pop())
+def _add_dependencies_to_stack(
+    import_stack: List[str], remaining_dependencies: Set[str]
+) -> None:
+    first_dependency = remaining_dependencies.pop()
 
-    return output_lib
+    if first_dependency in import_stack:
+        raise Exception("Circular import in yaml libraries")
```

Comment:
In practice circular imports are not a blocking issue because models only depend on ports, and ports don't depend on other ports.
So it would work to first create ports of all libraries in a first step, and then resolve all models of all libraries in a second step.

But ok to leave it this way

Reply:
I agree, however I find it cleaner to avoid circular import (maybe in the future we could define model extensions, or having other feature for which circular imports are blocking). And also I do not want to spend too much time now changing the logic of imports, will do it only if it is needed to replicate the C++ behavior.